File: py/common/primes.py
#!/usr/bin/env python3

"""primes.py

Functions for finding and identifying prime numbers.
"""

__author__ = 'Curtis Belmonte'

import math
from typing import Optional, Sequence

import common.arrays as arrs
import common.sequences as seqs


# Currently computed prime number terms (in sorted order)
_prime_sequence = [2]


def _compute_primes(n: int) -> None:
    """Precomputes and stores at least the first n prime numbers."""
    
    prime_count = len(_prime_sequence)

    # have the first n primes already been computed?
    if n < prime_count:
        return

    # TODO: implement incremental sieve?

    # set estimate and increment values according to PNT
    approx_gap = _estimate_prime_gap(n)
    estimate = 100 if n <= 25 else int(n * approx_gap)
    increment = int(approx_gap)

    # compute primes up to estimate, then step forward until n are found
    while len(_prime_sequence) < n:
        _compute_primes_up_to(estimate)
        estimate += increment


def _compute_primes_up_to(n: int) -> None:
    """Precomputes and stores the prime numbers up to n."""

    # have the numbers up to n already been computed?
    prime_max = _prime_sequence[-1]
    if prime_max >= n:
        return

    # prepare sieve of Eratosthenes for numbers prime_max + 1 to n
    sieve_size = n - prime_max
    sieve = [True] * sieve_size

    # sift out composite numbers using previously computed primes
    prime_count = len(_prime_sequence)
    for i in range(prime_count):
        rho = _prime_sequence[i]
        rho_start = seqs.next_multiple(rho, prime_max + 1)
        for j in range(rho_start, n + 1, rho):
            sieve[j - prime_max - 1] = False

    # sift out remaining composite numbers with newly found primes
    for i in range(sieve_size):
        if sieve[i]:
            rho = i + prime_max + 1
            _prime_sequence.append(rho)
            for j in range(rho**2 - prime_max - 1, sieve_size, rho):
                sieve[j] = False


def _estimate_prime_gap(n: int) -> float:
    """Returns an estimate for the average gap between the first n primes.

    Formula is based on a result derived from the Prime Number Theorem:
    https://en.wikipedia.org/wiki/Prime_number_theorem
    """

    log_n = math.log(n)
    log_log_n = math.log(log_n)
    log_n_sqr = log_n**2
    log_log_n_sqr = log_log_n**2
    return (log_n + log_log_n - 1 + (log_log_n - 2)/log_n
            - (log_log_n_sqr - 6 * log_log_n + 11)/(2 * log_n_sqr)
            + math.exp(1)/log_n_sqr)


def count_prime_factors(
        n: int,
        prime_nums: Optional[Sequence[int]] = None) -> int:

    """Returns the number of distinct prime factors of the natural number n.

    If provided, prime_nums must be an ordered sequence of prime numbers that
    contains at least the prime factors of n. If prime_factors is None, it will
    be calculated by this function.
    """

    # generate list of primes up to n if none given
    if prime_nums is None:
        prime_nums = primes_up_to(n)

    # check if n is prime to avoid worst-case performance
    if arrs.binary_search(prime_nums, n) is not None:
        factor_count = 1
    else:
        factor_count = 0
        for prime_num in prime_nums:
            # have all prime factors of n been found?
            if n == 1:
                break

            # if prime divides n, increment count and divide it out of n
            if n % prime_num == 0:
                factor_count += 1
                while n % prime_num == 0:
                    n //= prime_num

    return factor_count


def is_prime(n: int) -> bool:
    """Determines if the natural number n is prime."""

    # simple test for small n: 2 and 3 are prime, but 1 is not
    if n <= 3:
        return n > 1

    # check if multiple of 2 or 3
    if n % 2 == 0 or n % 3 == 0:
        return False

    # search for subsequent prime factors around multiples of 6
    max_factor = int(math.sqrt(n))
    for i in range(5, max_factor + 1, 6):
        if n % i == 0 or n % (i + 2) == 0:
            return False
    return True


def prime_factorization(n: int) -> Sequence[Sequence[int]]:
    """Computes the prime factorization of the natural number n.
    
    Returns a list of base-exponent pairs containing each prime factor and
    its power in the prime factorization of n.
    """

    d = 2
    factorization = []
    while d <= int(math.sqrt(n)):
        # compute power of i in factorization
        factor = [d, 0]
        div, mod = divmod(n, d)
        while mod == 0:
            n = div
            factor[1] += 1
            div, mod = divmod(n, d)

        # add factor to factorization if necessary
        if factor[1] > 0:
            factorization.append(factor)

        d += 1

    # no more prime factors above sqrt(n)
    if n > 1:
        factorization.append([n, 1])

    return factorization


def prime(n: int) -> int:
    """Returns the nth prime number."""
    _compute_primes(n)
    return _prime_sequence[n - 1]


def primes(n: int) -> Sequence[int]:
    """Returns the first n prime numbers in sorted order."""
    _compute_primes(n)
    return _prime_sequence[:n]


def primes_up_to(n: int) -> Sequence[int]:
    """Returns a sequence of the prime numbers <= n in sorted order."""

    _compute_primes_up_to(n)

    # return whole sequence if all primes <= n
    if _prime_sequence[-1] <= n:
        return _prime_sequence[:]

    # find the index of the last prime <= n
    i = 0
    while _prime_sequence[i] <= n:
        i += 1

    return _prime_sequence[:i]


def primorials(n: int) -> Sequence[int]:
    """Returns the first n primorial numbers in sorted order.

    The result is a sequence of n integers where each kth term is the product
    of the first k prime numbers, for 1 <= k <= n.
    """
    _compute_primes(n)
    return arrs.cumulative_products(_prime_sequence[:n])

File: jl/src/Primes.jl
#!/usr/bin/env julia

"""
Methods for finding and identifying prime numbers.
"""
module Primes

include("Sequences.jl")

# Currently computed prime number terms (in sorted order)
_prime_sequence = Int128[2]

"""
Precomputes and stores at least the first `n` prime numbers.
"""
function _compute_primes(n::Integer)
    global _prime_sequence

    # check if first n primes have already been computed
    if n < length(_prime_sequence)
        return
    end

    if n ≤ 25
        _compute_primes_up_to(97)
    else
        # set estimate and increment values according to PNT
        approx_gap = _estimate_prime_gap(n)
        estimate = trunc(BigInt, n * approx_gap)
        increment = trunc(BigInt, approx_gap)

        # get primes up to estimate, then step forward until n are found
        _compute_primes_up_to(estimate)
        while length(_prime_sequence) < n
            estimate += increment
            _compute_primes_up_to(estimate)
        end
    end
end

"""
Precomputes and stores the prime numbers up to `n`.
"""
function _compute_primes_up_to(n::Integer)
    global _prime_sequence

    # check if primes up to n have already been computed
    prime_max = last(_prime_sequence)
    if prime_max ≥ n
        return
    end

    # prepare sieve for numbers prime_max + 1 to n
    sieve_size = n - prime_max
    sieve = repeat([true], sieve_size)

    # sift out composite numbers using previously computed primes
    prime_count = length(_prime_sequence)
    for p in _prime_sequence
        p_start = Sequences.next_multiple(p, prime_max + 1)
        for j in p_start:p:n
            sieve[j - prime_max] = false
        end
    end

    # sift out composite numbers using previously computed primes
    for i in 1:sieve_size
        if sieve[i]
            p = i + prime_max
            push!(_prime_sequence, p)
            for j in (p^2 - prime_max):p:sieve_size
                sieve[j] = false
            end
        end
    end
end

"""
Returns an estimate for the average gap between the first `n` primes.

This formula is based on a result derived from the Prime Number Theorem:
https://en.wikipedia.org/wiki/Prime_number_theorem
"""
function _estimate_prime_gap(n::Integer)
    log_n = log(n)
    loglog_n = log(log_n)
    log_n2 = log(n^2)
    loglog_n2 = log(log_n^2)
    return (
        log_n + loglog_n - 1 + (loglog_n - 2)/log_n
        - (loglog_n2 - 6loglog_n + 11)/(2log_n2) + exp(1)/log_n2
    )
end

"""
Resets the currently cached list of prime numbers.
"""
function _reset_prime_cache()
    global _prime_sequence
    _prime_sequence = Int128[2]
    return
end

"""
    prime(n::Integer)

Finds the `n`th prime number.

## Preconditions
- `n > 0`

## Examples
```jldoctest
julia> prime(1)
2

julia> prime(6)
13
```
"""
function prime(n::Integer)
    _compute_primes(n)
    global _prime_sequence
    return _prime_sequence[n]
end

"""
    prime_factorization(n::Integer)

Computes the prime factorization of the natural number `n`.

Result is a 1D array of base-exponent tuple pairs, containing each prime factor
and its power in the prime factorization of n.

## Preconditions
- `n ≥ 2`

## Examples
```jldoctest
julia> prime_factorization(UInt8(2))
1-element Array{Tuple{UInt8,UInt8},1}:
 (0x02, 0x01)

julia> prime_factorization(Int64(12))
2-element Array{Tuple{Int64,Int64},1}:
 (2, 2)
 (3, 1)

julia> prime_factorization(BigInt(600))
3-element Array{Tuple{BigInt,BigInt},1}:
 (2, 3)
 (3, 1)
 (5, 2)
```
"""
function prime_factorization(n::Integer)
    n_type = typeof(n)

    # populate array with (factor, power) pairs
    factor = 2 * oneunit(n_type)
    factorization = Array{Tuple{n_type, n_type}, 1}()
    while factor ≤ floor(n_type, sqrt(n))
        # compute power of factor in factorization
        power = zero(n_type)
        (d, r) = divrem(n, factor)
        while r == 0
            n = d
            power += 1
            (d, r) = divrem(n, factor)
        end

        # add (factor, power) pair to factorization
        if power > zero(n_type)
            push!(factorization, (factor, power))
        end

        factor += 1
    end

    # append remaining prime factor if necessary
    if n > 1
        push!(factorization, (n, 1))
    end

    return factorization
end

end
